block-commits/pullRequestChecker.js

```diff
@@ -0,0 +1,94 @@
+const {info} = require('@actions/core');
+const {
+  context,
+  getOctokit,
+} = require('@actions/github');
+
+class PullRequestChecker {
+  constructor(repoToken, actionMerge, actionFixup) {
+    this.client = getOctokit(repoToken);
+    this.actionMerge = actionMerge;
+    this.actionFixup = actionFixup;
+  }
+
+  async process() {
+    const listArgs = {
+      ...context.repo,
+      pull_number: context.issue.number,
+      per_page: 100,
+    };
+    const commitsEndpoint = this.client.rest.pulls.listCommits.endpoint.merge(listArgs);
+    const commits = await this.client.paginate(commitsEndpoint);
+
+    info(`Action for fixup commits: ${this.actionFixup}`);
+    info(`Action for merge commits: ${this.actionMerge}`);
+    info(`${commits.length} commit(s) in the pull request`);
+
+    let autosquashCommits = 0;
+    let mergeCommits = 0;
+    for (const {commit: {message}, sha, url, parents} of commits) {
+      const isAutosquash = message.startsWith('fixup!') || message.startsWith('squash!');
+      const isMergeCommit = parents.length > 1;
+
+      if (this.actionFixup != 'none' && isAutosquash) {
+        info(`Commit ${sha} is an autosquash commit: ${url}`);
+
+        autosquashCommits++;
+      }
+      if (this.actionMerge != 'none' && isMergeCommit) {
+        info(`Commit ${sha} is an merge commit: ${url}`);
+
+        mergeCommits++;
+      }
+    }
+    const reviewMessages = [];
+    const failMessages = [];
+    if (autosquashCommits) {
+      const message = `${autosquashCommits} commit(s) that need to be squashed`;
+      if (this.actionFixup == 'request-changes') {
+        reviewMessages.push(message);
+      }
+      if (this.actionFixup == 'fail') {
+        failMessages.push(message);
+      }
+    }
+    if (mergeCommits) {
+      const message = `${mergeCommits} merge commits`;
+      if (this.actionMerge == 'request-changes') {
+        reviewMessages.push(message);
+      }
+      if (this.actionMerge == 'fail') {
+        failMessages.push(message);
+      }
+    }
+    if (this.actionFixup == 'request-changes' || this.actionMerge == 'request-changes') {
+      const okMessage = 'PR no longer has any merge or fixup commits.';
+      const failPrefix = 'PR requires a rebase. Found: ';
```

Comment:
We should be careful requiring rebase.
whenever we do so, we should instruct to use
```
git rebase -i $(git merge-base master head)
```

Reply:
Would you like me to add that to the message? Or can you suggest a better phrasing to avoid suggesting a rebase?

Reply:
For merge commits the current wording is fine.
for fixups you can omit that for now. we can improve the wording once we start using this and can see this in practice

Reply:
I reworded both the ok and fail messages. PTAL.